test/NuGet.Core.Tests/NuGet.Packaging.Test/ContentModelTests/PatternExpressionMatchTests.cs

```diff
@@ -0,0 +1,198 @@
+            Assert.NotNull(match_result);
+            object result;
+            match_result.Properties.TryGetValue("tfm_raw", out result);
+            Assert.Equal(expected_result, result);
+        }
+
+        [Fact]
+        public void TokenSegment_multipleTokens()
+        {
+            Dictionary<string, ContentPropertyDefinition> propertyDefinitions =
+                new Dictionary<string, ContentPropertyDefinition>() {
+        {"name", new ContentPropertyDefinition("any", parser: (o, t) => o) },
+        {"version", new ContentPropertyDefinition("any", parser: (o, t) => o) }
+            };
+            var pattern = new PatternDefinition("contentItem/{name}/{version}");
+            var expression = new PatternExpression(pattern);
+            var path = "contentItem/mypackage/1.0.0";
+            var expected_name = "mypackage";
+            var expected_version = "1.0.0";
+            var match_result = expression.Match(path, propertyDefinitions);
+            Assert.NotNull(match_result);
+            object name_result;
+            object version_result;
+            match_result.Properties.TryGetValue("name", out name_result);
+            match_result.Properties.TryGetValue("version", out version_result);
+            Assert.Equal(expected_name, name_result);
+            Assert.Equal(expected_version, version_result);
+        }
+
+        [Fact]
+        public void TokenSegment_customDelimiter()
+        {
+            Dictionary<string, ContentPropertyDefinition> propertyDefinitions =
+                new Dictionary<string, ContentPropertyDefinition>() {
+        {"name", new ContentPropertyDefinition("any", parser: (o, t) => o) },
+        {"version", new ContentPropertyDefinition("any", parser: (o, t) => o) }
+            };
+            var pattern = new PatternDefinition("contentItem/{name}--{version}");
+            var expression = new PatternExpression(pattern);
+            var path = "contentItem/mypackage--1.0.0/file.dll";
+            var expected_result = "mypackage";
+            var match_result = expression.Match(path, propertyDefinitions);
+            Assert.NotNull(match_result);
+            object result;
+            match_result.Properties.TryGetValue("name", out result);
+            Assert.Equal(expected_result, result);
+        }
+
+        [Fact]
+        public void TokenSegment_tokenNotFound()
+        {
+            Dictionary<string, ContentPropertyDefinition> propertyDefinitions =
+                new Dictionary<string, ContentPropertyDefinition>()
+                {
+                    // Empty propertyDefinitions, no token definition available
+                };
+            var pattern = new PatternDefinition("contentItem/{name}");
+            var expression = new PatternExpression(pattern);
+            var path = "contentItem/mypackage/file.txt";
+            Assert.Throws<Exception>(() => expression.Match(path, propertyDefinitions));
```

Comment:
Be careful about asserting that something throws.  In the simple case (this), you aren't verifying anything beyond the throw.  It would be a stronger verification to assert that the thrown exception (e.g.:  type, parameter name, message, etc.) is what you expect.  In this case, [the exception type is `System.Exception`](https://github.com/NuGet/NuGet.Client/blob/23f51111dc157ce393c9dc96fb87450ec2aec676/src/NuGet.Core/NuGet.Packaging/ContentModel/Infrastructure/Parser.cs#L173), but you can still verify the message:

```suggestion
            Exception exception = Assert.Throws<Exception>(() => expression.Match(path, propertyDefinitions));
            Assert.Equal("Unable to find property definition for {name}", exception.Message);
```

Apply this feedback wherever you call `Assert.Throws<T>(...)`

Reply:
I dont' think it's a good idea to assert exception text messages. [It causes barriers to people who use locales other than English from contributing to our repo](https://github.com/NuGet/NuGet.Client/pull/5002#issuecomment-1679707317)

Reply:
I think we agree that the following are desirable:

* eliminating barriers to others to contribute
* improving test quality

In the context of this PR, until proven otherwise, I don't know that the two are mutually exclusive.

In the general case, it wasn't clear to me what the root problems were in the comment you linked.  Can you summarize here?

In most product code, we `string.Format(CultureInfo.CurrentCulture, ...)` to produce a culture-specific formatted message.  If we are following all localization guidance, why wouldn't a test that checks the message succeed in non-en-US locales?  My guess is that we aren't following all localization guidance, either in the product or in the test or both.  For example, the product uses `string.Format(CultureInfo.CurrentCulture, ...)` with a localizable resource string but a test uses a hardcoded en-US string for the expected result.  I would expect that to fail for non-en-US locales.  This is a test bug, and it should be fixed, so users don't have to switch locales to run tests.

There are a couple problems related to this test:

1.  The product code throws `System.Exception`, which is discouraged, but it is what it is.  It would be better if it threw a `ParserException` exception with a property that indicates the problematic token, much like `ArgumentException.ParamName`.  However, changing the exception type is out of scope for this PR and would change the behavior of a [public API](https://github.com/NuGet/NuGet.Client/blob/23f51111dc157ce393c9dc96fb87450ec2aec676/src/NuGet.Core/NuGet.Packaging/PublicAPI/netcoreapp5.0/PublicAPI.Shipped.txt#L67).  (I suspect throwable exceptions are outside the purview of the API analyzer, but I don't really know.)
2.  The thrown exception doesn't use a resource string.  This isn't localizable.

If the test formats the string like the product, would that fail on non-en-US locales?

```C#
Assert.Equal(string.Format(CultureInfo.CurrentCulture, "Unable to find property definition for {{{0}}}", "{name}"), exception.Message);
```

Reply:
Looking at the first comment more carefully, I disagree with this:

> In the simple case (this), you aren't verifying anything beyond the throw. It would be a stronger verification to assert that the thrown exception (e.g.: type, parameter name, message, etc.) is what you expect.

Automated tests greatest value is catching regressions. Do you believe that changing an error message is a regression? I don't. Therefore asserting a message has a very specific string doesn't help catch logic errors, but it does slow down changes improving error messages, and therefore meets my definition of technical debt (slows us down without providing value).

I've worked on one project that had a LOT of unit tests asserting all the implementation details that the team could think of for most classes in the app. This means that every time you add a new feature and make minor changes to existing code, or fix a bug, there's a bunch of failing tests because a bunch of internal state that is irrelevant to customer experience was being asserted, slowing down team velocity. It's well intended, no doubt, but unfortunately counter-productive. Refactoring was an absolute nightmare as we effectively just deleted the existing tests, refactored the code, then wrote new tests, in which case the tests were useless because hey didn't help catch regressions. I think the most valuable tests validate contracts, not implementation details.

So, I don't agree that we should have stronger verification of error messages. Here, the important thing is that an error is raised when the invalid input is provided. In my opinion the exact text of the error message is not important.

I agree that it would be better if the exception was `ArgumentException`, so we could assert that `.ParamName == "whatever"`. If the current exception message has the param name (and expanding to general cases, not just this exception message, if the error message contains a value that was passed in as a method parameter), then I'm fine with checking that `exception.Message.Should().Contain("paramName")`. But I need convincing that asserting exact exception message text is valuable, rather than tech debt.

> In most product code, we string.Format(CultureInfo.CurrentCulture, ...) to produce a culture-specific formatted message. If we are following all localization guidance, why wouldn't a test that checks the message succeed in non-en-US locales?

If the test is running on a binary that contains localization, then any `StringResource.Whatever` will return the most appropriate string for the current locale, but then a test that validates `string.Equals(actualString, "Expected string in English")` is obviously going to fail for non-English locales.

Additionally, the `CultureInfo` passed to `string.Foramt` only applies culture formatting to numbers and dates. Maybe more, but certainly not string resources obtained from resx files. Side note, PM UI's package published date is the only place in the entire codebase where we actually use culture-aware formatting. I don't know of a single place where we use format to provide culture aware to format numbers (thousands separator, whole numbers from fraction). So passing a `CultureInfo` to `string.Format` proabbly doesn't provide NuGet.Client with any benefit, since the one place we format dates is in XAML, not C# code. But someone enabled a code analysis warning that makes sense in other products, so that's why we have `string.Format(CultureInfo.CurrentCulture,` littered throughout our repo. I didn't argue when it was being added because I need to pick my battles, and it isn't harmful, only adds boilerplate.

Anyway, in case you're not familiar with resx string localization, have a quick look at [one of the `.Designer.cs` files](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Common/Strings.Designer.cs). The class has [a private static CultureInfo](https://github.com/NuGet/NuGet.Client/blob/4d815bf71aa1ce5fe8d9d9da6573303d7093c2f1/src/NuGet.Core/NuGet.Common/Strings.Designer.cs#L29), and [every string resource property passes it to `ResourceManager.GetString`](https://github.com/NuGet/NuGet.Client/blob/4d815bf71aa1ce5fe8d9d9da6573303d7093c2f1/src/NuGet.Core/NuGet.Common/Strings.Designer.cs#L68).  Since the usual pattern for using parameterized string resources is `string.Format(someCulture, Strings.Error_Something, arg1, arg2)`, this means that the string is retrieved from the resource manager BEFORE being passed to `string.Format`, so the culture passed to `string.Format` cannot affect the string resource localization used.

> However, changing the exception type is out of scope for this PR and would change the behavior of a public API. (I suspect throwable exceptions are outside the purview of the API analyzer, but I don't really know.)

You're right, the public API analyzer doesn't track what exceptions are being thrown, so changing exceptions wouldn't be caught by tooling.

And changing from `Exception` to `ArgumentException` would only break someone doing `exception.GetType() == typeof(Exception)`, or equivalent. But I hope you agree that doesn't sound like a very valuable thing to do outside this specific test (and a test validating the contract of which exception type is thrown seems like a very valid test to me). Customers using the NuGet CLient SDK and doing `catch (Exception)` will continue to catch the exception if we change the exception type. However, if we were later to change it from `ArgumentException` to anything else (that does not extend `ArgumentException`) then that would be a breaking change with "realistic" risk.

Having written that, `PatternExpression` doesn't sound like a useful API for customers using the NuGet Client SDK in their own app. It sounds like something that should be internal, an implementation detail of some package related API. However, we have no way to know if anyone actually uses it. But since it sounds like a "useless" API outside of NuGet.Packaging.dll, I'd be reasonably open to making a breaking change in this type, but I don't know if other people in the team feel the same. We have had failed insertions in the past because of broken APIs, and other times where insertion didn't catch the break, and another product team had to change their code when NuGet flowed into their repo (or a feature without automated tests failed, but these days the .NET SDK has a LOT of tests)

Reply:
@Nigusu-Allehu, @zivkan and I chatted offline and agreed on the following:

* Change [this exception](https://github.com/NuGet/NuGet.Client/blob/23f51111dc157ce393c9dc96fb87450ec2aec676/src/NuGet.Core/NuGet.Packaging/ContentModel/Infrastructure/Parser.cs#L173) to `System.ArgumentException`.
* Pass the current message to `ArgumentException'`s constructor for the `message` argument and the token for the `paramName` argument.  Here the token is `"{name}"`.
* The test should assert that `ArgumentException` is thrown and that `ArgumentException.ParamName` == `"{name}"`.

The spirit of the feedback is that a negative test should try to verify both failure (here, that an exception is thrown) and that the failure reason is expected (here, that the exception is the one you expect).